raft/src/main/java/org/apache/kafka/raft/CandidateState.java

```diff
@@ -235,6 +240,12 @@ public int epoch() {
         return highWatermark;
     }
 
+    @Override
+    public boolean canGrantVote(int candidateId, boolean isLogUpToDate) {
```

Comment:
Might be worth a comment about the behavior when `candidateId` is equal to `localId`. Although the candidate votes for itself, this vote is implicit and not "granted."